pkg/kubelet/kuberuntime/kuberuntime_logs.go

```diff
@@ -0,0 +1,332 @@
+	"github.com/fsnotify/fsnotify"
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// Notice that the current kuberuntime logs implementation doesn't handle
+// log rotation.
+// * It will not retrieve logs in rotated log file.
+// * If log rotation happens when following the log:
+//   * If the rotation is using create mode, we'll still follow the old file.
+//   * If the rotation is using copytruncate, we'll be reading at the original position and get nothing.
+// TODO(random-liu): Support log rotation.
+
+// streamType is the type of the stream.
+type streamType string
+
+const (
+	stderrType streamType = "stderr"
+	stdoutType streamType = "stdout"
+
+	// timeFormat is the time format used in the log.
+	timeFormat = time.RFC3339Nano
+	// blockSize is the block size used in tail.
+	blockSize = 1024
+)
+
+var (
+	// eol is the end-of-line sign in the log.
+	eol = []byte{'\n'}
+	// delimiter is the delimiter for timestamp and streamtype in log line.
+	delimiter = []byte{' '}
+)
+
+// logMessage is the internal log type.
+type logMessage struct {
+	timestamp time.Time
+	stream    streamType
+	log       []byte
+}
+
+// reset resets the log to nil.
+func (l *logMessage) reset() {
+	l.timestamp = time.Time{}
+	l.stream = ""
+	l.log = nil
+}
+
+// logOptions is the internal type of all log options.
+type logOptions struct {
+	tail      int64
+	since     time.Time
+	follow    bool
+	timestamp bool
+}
+
+// newLogOptions convert the api.PodLogOptions to internal logOptions.
+func newLogOptions(apiOpts *api.PodLogOptions, now time.Time) *logOptions {
+	opts := &logOptions{
+		tail:      -1, // -1 by default which means read all logs.
+		follow:    apiOpts.Follow,
+		timestamp: apiOpts.Timestamps,
+	}
+	if apiOpts.TailLines != nil {
+		opts.tail = *apiOpts.TailLines
+	}
+	if apiOpts.SinceSeconds != nil {
+		opts.since = now.Add(-time.Duration(*apiOpts.SinceSeconds) * time.Second)
+	}
+	if apiOpts.SinceTime != nil && apiOpts.SinceTime.After(opts.since) {
+		opts.since = apiOpts.SinceTime.Time
+	}
+	return opts
+}
+
+// ReadLogs read the container log and redirect into stdout and stderr.
+func ReadLogs(path string, apiOpts *api.PodLogOptions, stdout, stderr io.Writer) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("failed to open log file %q: %v", path, err)
+	}
+	defer f.Close()
+
+	// Search start point based on tail line.
+	opts := newLogOptions(apiOpts, time.Now())
+	start, err := tail(f, opts.tail)
+	if err != nil {
+		return fmt.Errorf("failed to tail %d lines of log file %q: %v", opts.tail, path, err)
+	}
+	if _, err := f.Seek(start, os.SEEK_SET); err != nil {
+		return fmt.Errorf("failed to seek %d in log file: %v", start, path)
+	}
+
+	// Start parsing the logs.
+	r := bufio.NewReader(f)
+	// Do not create watcher here because it is not needed if `Follow` is false.
+	var w *fsnotify.Watcher
+	var parse parseFunc
+	msg := &logMessage{}
+	for {
+		l, err := r.ReadBytes(eol[0])
```

Comment:
is there a way to specify a multi-character delimiter?  its probably not needed for stdout/stderr but for the future state of handling more arbitrary log files (stacktraces in an application log file jump to mind) it'd be nice to have multi-char event delimiters sorted out already.

if this is out of scope for now, please ignore.